### PR TITLE
Adds LIGHT modes and capabilities (passive floats, improved sampling)

### PR DESCRIPTION
**Pull request by pwolfram**
This merge adds passive particle floats, generalizes and improves LIGHT code, and adds support for horizontal interpolation of sampling for salinity, temperature, and BGC quantities.

It also fixes the condition where partial bottom cells allows passive floats to advect out of the wetted domain.  The solution, following consultation with @maltrud, is to simply ensure that particles are contained in the viable range (ssh - dz/2, bottomDepth - dz/2), which essentially returns particles to the nearest wet cell center vertically.

### CODE REVIEW

**Comment by pwolfram** [comment]
cc @bradyrx, @xylar as we have discussed 3D passive float particle tracking in the past.

**Comment by pwolfram** [comment]
cc @sbrus89 as an example of an MPAS-O PR that we will use for testing.

**Comment by pwolfram** [comment]
Particle setup files are at https://gist.github.com/pwolfram/65266db0e28ba9268a88cd88376580ce

**Comment by mark-petersen** [comment]
@pwolfram please confirm that this PR passes the nightly test suite.  Only the cases with LIGHT are relevant, but it's easy to test the whole thing.

File: src/core_ocean/analysis_members/mpas_ocn_lagrangian_particle_tracking.F [hunk 1]
```diff
@@ -705,44 +712,73 @@ subroutine ocn_compute_lagrangian_particle_tracking(domain, timeLevel, err)!{{{
...
-            particlePosition, zLevelParticle, zMid, zTop, vertVelocityTop, xVertex, yVertex, zVertex, &
-            meshPool, areaBArray)
+          !verify particles are still in domain !{{{
+          ! check and verify that free floating particle hasn't gone out of boundary
+          if (iLevel == -1) then
+            call mpas_log_write(' Warning: Particle $i location of $r is higher than zMid of $r', realArgs=(/ zLevelParticle, zMid(1, iCell) /), intArgs=(/iCell/))
+          end if
+          if (iLevel == 0) then
+            call mpas_log_write(' Warning: Particle $i location of $r is deeper than zMid at bottom $r', realArgs=(/ zLevelParticle, zMid(maxLevelCell(iCell),iCell) /), intArgs=(/iCell/))
+          end if
```
Comment (pwolfram) [inline src/core_ocean/analysis_members/mpas_ocn_lagrangian_particle_tracking.F]: @bradyrx, this is an example of code that probably needs commented out (above) as we teach you about PRs.

Comment (pwolfram) [reply src/core_ocean/analysis_members/mpas_ocn_lagrangian_particle_tracking.F]: It is now wrapped in a debug statement, so this should be a non-issue.

File: testing_and_setup/compass/utility_scripts/LIGHTparticles/make_particle_file.py [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+        return val
+    else:
+        return defaults[name]
+
+def ensure_shape(start, new):
+    if isinstance(new, (int, float)):
+        new = new*np.ones_like(start)
+    return new
+
+def southern_ocean_only_xyz(x, y, z, maxNorth=-45.0):
```
Comment (pwolfram) [inline testing_and_setup/compass/utility_scripts/LIGHTparticles/make_particle_file.py]: @bradyrx, I've updated this script to do the filtering via something like `./make_particle_file.py -i /home/ccsm-data/inputdata/ocn/mpas-o/oQU240/ocean.QU.240km.151209.nc -g /home/ccsm-data/inputdata/ocn/mpas-o/oQU240/mpas-o.graph.info.151209.part. -p 72 -o /lcrc/group/acme/pwolfram/acme_scratch/anvil/20180626.GMPAS-IAF.T62_oQU240.anvil.particles/run/particles.nc --spatialfilter SouthernOceanXYZ -t ['passive','surface']` (on blues for me).

Note, I've "hard-coded" the option `maxNorth=-45.0` for now.  I tested by completing a quick G-case 240km run over 30 days.

File: src/core_ocean/analysis_members/Registry_lagrangian_particle_tracking.xml [hunk 3]
```diff
@@ -307,6 +315,13 @@
 		<var name="transfered" type ="integer" dimensions="nParticles Time" units="unitless"
 			 description="flag to monitor if the particle was transfered"
 		/>
+<!-- TEMP FIELDS THAT NEED MOVED TO NonHalo below after verifying interpolation is correct -->
+		<var name="temperature" type="real" dimensions="nParticles Time" units="kg m^{-3}"
```
Comment (bradyrx) [inline src/core_ocean/analysis_members/Registry_lagrangian_particle_tracking.xml]: @pwolfram -- should temperature units be "degC" or "K"?

Comment (pwolfram) [reply src/core_ocean/analysis_members/Registry_lagrangian_particle_tracking.xml]: It should be, thanks for catching this typo.  It should also have in its description that it is potential temperature, not just regular temperature.

**Comment by pwolfram** [comment]
Verified that LIGHT is BFB against `fd71e645` using SOMA test case for "spot checking" (the case that failed in https://github.com/pwolfram/MPAS-Model/pull/21 to find the bug fixed by `
e4f2dc5 `.

**Comment by mark-petersen** [comment]
It also failed this one:
```
 ** Running case ZISO 20km - Smoke Test
   ** FAIL (See case_outputs/ZISO_20km_-_Smoke_Test for more information)
```
because of this line:
```
testing_and_setup/compass/utility_scripts/LIGHTparticles/make_particle_file.py:
  from pyamg.classical import split, interpolate as amginterp
```
I tried to install pyamg in my local python installation but couldn't get it to work. I'm sure I just installed it wrong.  If you can get that ZISO smoke test case to run successfully, go ahead and merge.  Then we can make a PR into E3SM.

File: testing_and_setup/compass/utility_scripts/LIGHTparticles/make_particle_file.py [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+
+Phillip J. Wolfram
+Last Modified: 08/03/2018
+"""
+
+import netCDF4
+import numpy as np
+from scipy import spatial
+from scipy import sparse
+from pyamg.classical import split, interpolate as amginterp
```
Comment (bradyrx) [inline testing_and_setup/compass/utility_scripts/LIGHTparticles/make_particle_file.py]: @pwolfram, line 19 should be wrapped in a try block per our call with ANL and @mark-petersen's error. This of course will break if the user tries the `--downsample` flag without `pyamg`, but that should be fine. For whatever reason, `pyamg` is not an easy install (took many tries for me).

Comment (mark-petersen) [reply testing_and_setup/compass/utility_scripts/LIGHTparticles/make_particle_file.py]: @bradyrx thanks for the suggestion.  I agree, put it in a try block, and have it actually fail only when the library is needed.  That will keep our nightly testing suite more accessible and portable across users and machines.

Comment (bradyrx) [reply testing_and_setup/compass/utility_scripts/LIGHTparticles/make_particle_file.py]: Ideally we eventually need a better solution than `pyamg` for coarsening, but I don't think it's crucial for this PR. It's a feasible solution for oQU240, but takes ~40min for 60to30 and ~20hr for 30to10.

Comment (bradyrx) [reply testing_and_setup/compass/utility_scripts/LIGHTparticles/make_particle_file.py]: Note my current hacky solution here: https://gist.github.com/bradyrx/ddb6c944e425ac5e63b8e8e7361809dd

I'm just running `make_particle_file.py` without downsampling, and then removing full columns with that gist by going through lat/lon with a fixed stride.

This approximately pulls columns at random due to the unstructured mesh indexing, but I don't suggest it as a formal method in ocean/develop.

Here, white dots are seeding at every grid cell and red are after running my script:

**Comment by pwolfram** [comment]
@mark-petersen, here is the expected output from regression testing:

```
┌─[pwolfram][weaklynonlinear][~/Documents/MPAS_development/LIGHT_test_refractorPR21/LIGHT_feature_test_brb/MPAS-Model/testing_and_setup/compass/light_candidate][14:54][±][LIGHT_clean_updates ✗]
└─▪ ./nightly_ocean_test_suite.py 
 ** Running case Periodic Planar 20km - LIGHT particle general test
   ** FAIL (See case_outputs/Periodic_Planar_20km_-_LIGHT_particle_general_test for more information)
 ** Running case Periodic Planar 20km - LIGHT particle region reset test
   ** FAIL (See case_outputs/Periodic_Planar_20km_-_LIGHT_particle_region_reset_test for more information)
 ** Running case Periodic Planar 20km - LIGHT particle time reset test
   ** FAIL (See case_outputs/Periodic_Planar_20km_-_LIGHT_particle_time_reset_test for more information)
 ** Running case SOMA 32km - LIGHT particle time general test
   ** FAIL (See case_outputs/SOMA_32km_-_LIGHT_particle_time_general_test for more information)
 ** Running case ZISO 20km - LIGHT particle time general test
      PASS
TEST RUNTIMES:
00:16 Periodic_Planar_20km_-_LIGHT_particle_general_test
00:15 Periodic_Planar_20km_-_LIGHT_particle_region_reset_test
00:22 Periodic_Planar_20km_-_LIGHT_particle_time_reset_test
02:00 SOMA_32km_-_LIGHT_particle_time_general_test
00:41 ZISO_20km_-_LIGHT_particle_time_general_test
Total runtime 03:34
┌─[pwolfram][weaklynonlinear][~/Documents/MPAS_development/LIGHT_test_refractorPR21/LIGHT_feature_test_brb/MPAS-Model/testing_and_setup/compass/light_candidate][14:57][±][LIGHT_clean_updates ✗]
└─▪ grep FAIL case_outputs/*
case_outputs/Periodic_Planar_20km_-_LIGHT_particle_general_test: ** FAIL Comparison of zLevelParticle between forward/analysis_members/lagrPartTrack.0001-01-01_00.00.00.nc and
case_outputs/Periodic_Planar_20km_-_LIGHT_particle_region_reset_test: ** FAIL Comparison of zLevelParticle between forward/analysis_members/lagrPartTrack.0001-01-01_00.00.00.nc and
case_outputs/Periodic_Planar_20km_-_LIGHT_particle_time_reset_test: ** FAIL Comparison of zLevelParticle between forward/analysis_members/lagrPartTrack.0001-01-01_00.00.00.nc and
case_outputs/SOMA_32km_-_LIGHT_particle_time_general_test: ** FAIL Comparison of particleTemperature between forward/analysis_members/lagrPartTrack.0001-01-01_00.00.00.nc and
case_outputs/SOMA_32km_-_LIGHT_particle_time_general_test: ** FAIL Comparison of particleSalinity between forward/analysis_members/lagrPartTrack.0001-01-01_00.00.00.nc and
```
The failure is due to redoing the `zLevelParticle` coordinate for passively-floating particles and temperature and salinity were added to SOMA, hence the failure since there was no previous baseline.

**Comment by pwolfram** [comment]
@maltrud, here is the latest version of LIGHT (ready to merge) suitable for use in the 30to10 case.

**Comment by pwolfram** [comment]
@bradyrx, I sorted out the issue, cleaned up this PR, and it should be ready for use and to merge.

Screen Shot 2019-04-24 at 3 05 45 PM
Screen Shot 2019-04-24 at 3 06 27 PM

These examples use passive particles (3D floating mode) and fixes the bug you found.

**Comment by bradyrx** [comment]
@pwolfram, great news. Thanks for putting the time into this. @maltrud -- are you good to cherry pick from this branch? Alternatively, I can checkout this PR/branch on grizzly and point you to the files to diff.

**Comment by maltrud** [comment]
@bradyrx , thanks for the offer.  i would really appreciate it if you just pointed me to your checkout.  i hate git.  i hate cherry-picking.  feels really good to say this on a git thread.

**Comment by mark-petersen** [comment]
Merged locally to ocean/develop, compiled with gnu and intel/debug.  Three comments:

1. Passes and bfb with ocean/develop except for LIGHT variables, as expected.  Only variables with diffs are:
```
grep FAIL case_outputs/Periodic_Planar_20km_-_LIGHT_particle_time_reset_test
 ** FAIL Comparison of zLevelParticle between forward/analysis_members/lagrPartTrack.0001-01-01_00.00.00.nc and
grep FAIL case_outputs/Periodic_Planar_20km_-_LIGHT_particle_region_reset_test
 ** FAIL Comparison of zLevelParticle between forward/analysis_members/lagrPartTrack.0001-01-01_00.00.00.nc and
```
and in `ocean/ziso/20km/default` I get
`ERROR: Field sizes don't match in different files.` for all LIGHT variables, which I assume is expected.

2. Note that `testing_and_setup/compass/utility_scripts//LIGHTparticles/make_particle_file.py` used in `ziso/20km/default/forward/` requires the library `pyamg`, which is not in e3sm unified. I tested it with my own python installation and got everything to work, but we should add pyamg to the compass environment (cc @xylar).

3. In debug mode there is a HUGE amount of text written to the screen. Much of it is large arrays or variable names, so I think it is unintentional. It needs to be removed, as we often run short tests in debug mode.  You can put it under a special preprocessing verbose flag if you want to keep it.
```
iCell=
 iLevel=
 particleVelocityVert=
 lambda=
 uVertex=
```
If you take all that out, I can finish my testing with debug on.

File: testing_and_setup/compass/ocean/soma/analysis/check_particle_sampling.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env python
```
Comment (xylar) [inline testing_and_setup/compass/ocean/soma/analysis/check_particle_sampling.py]: please check that this script is python 3 compatible.

File: testing_and_setup/compass/utility_scripts/LIGHTparticles/animated_particles.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
+#! /usr/bin/env python
```
Comment (xylar) [inline testing_and_setup/compass/utility_scripts/LIGHTparticles/animated_particles.py]: please check that this script is python 3 compatible.

File: testing_and_setup/compass/utility_scripts/LIGHTparticles/make_particle_file.py [hunk 7]
```diff
@@ -0,0 +1,490 @@
+#!/usr/bin/env python
```
Comment (xylar) [inline testing_and_setup/compass/utility_scripts/LIGHTparticles/make_particle_file.py]: please check that this script is python 3 compatible.

**Comment by pwolfram** [comment]
@mark-petersen, would you mind pointing me to your output for 1?  I am not sure what to think of   
```
and in ocean/ziso/20km/default I get
ERROR: Field sizes don't match in different files. for all LIGHT variables, which I assume is expected.
```
and am thinking the baseline for LIGHT may be outdated.  What commit is the base for the baseline?

On 3, I'll make those changes so debug is cleaner.  I think it will be a simple fix and will let you know once I push the change.

**Comment by pwolfram** [comment]
@xylar, the request for python3 compatibility is scope creep given that the current repo is not python3 compatible.  I do not expect you to update these scripts, but this is not the right place to update them to python3 because they are already tested and fully within the workflow.  To update them now is additional work that will prevent the needs of this PR from being immediately met.  I'm happy to submit an issue detailing the update to python3 for this code once this is merged but disagree we need to do this now, especially since we don't have a "trusted" python environment that can be used to test against with all libraries, e.g., pyamg.  Long term, I think we need this for the regression suite and would argue that we need a pointer in the MPAS repo to this as a submodule or some such to make sure the python repo and the MPAS repos are synced.  Again, I would argue this is completely out of scope for this PR, but it certainly warrants a follow-on issue / PR.

**Comment by xylar** [comment]
Alright, I'll reluctantly drop my concern.  This PR was put in nearly a year ago.  I can see that testing python 3 is potentially a little tricky.  But each new python script we add complicates the conversion to python 3 just a bit more.  I want to impress on everyone that I really don't think any new python 2.7 code should be being written and a concerted effort is needed to port our existing scripts.

I don't think a separate GitHub issue for these particular scripts is useful.  Please do put it on your own to-do list to convert these scripts and others related to your test cases ASAP.

**Comment by pwolfram** [comment]
@mark-petersen, I've pushed a commit that should handle your concern with spurious debug messages.  The PR should be ready for your review again.

**Comment by pwolfram** [comment]
@mark-petersen, expected errors are:

```
└─▪ grep 'ERROR\|FAIL' case_outputs/*
case_outputs/Periodic_Planar_20km_-_LIGHT_particle_general_test: ** FAIL Comparison of zLevelParticle between forward/analysis_members/lagrPartTrack.0001-01-01_00.00.00.nc and
case_outputs/Periodic_Planar_20km_-_LIGHT_particle_region_reset_test: ** FAIL Comparison of zLevelParticle between forward/analysis_members/lagrPartTrack.0001-01-01_00.00.00.nc and
case_outputs/Periodic_Planar_20km_-_LIGHT_particle_time_reset_test: ** FAIL Comparison of zLevelParticle between forward/analysis_members/lagrPartTrack.0001-01-01_00.00.00.nc and
case_outputs/SOMA_32km_-_LIGHT_particle_time_general_test:ERROR: Field 'particleTemperature' does not exist in both
case_outputs/SOMA_32km_-_LIGHT_particle_time_general_test: ** FAIL Comparison of particleTemperature between forward/analysis_members/lagrPartTrack.0001-01-01_00.00.00.nc and
case_outputs/SOMA_32km_-_LIGHT_particle_time_general_test:ERROR: Field 'particleSalinity' does not exist in both
case_outputs/SOMA_32km_-_LIGHT_particle_time_general_test: ** FAIL Comparison of particleSalinity between forward/analysis_members/lagrPartTrack.0001-01-01_00.00.00.nc and
```

other errors are a cause for concern and need investigation.

**Comment by bradyrx** [comment]
@pwolfram, I think your `make_particle_file.py` script is outdated in this PR as well. Recall that you updated it to manually coarsen to avoid the poor scaling on `pyamg`. Here is a gist with the most recent version with downsampling you made: https://gist.github.com/bradyrx/2d115df15e6d76b453ed75e5eb4183c8

**Comment by pwolfram** [comment]
@mark-petersen, @bradyrx is right-- I'm missing one commit on here that needs to be included from the other branch.

**Comment by pwolfram** [comment]
@bradyrx and @mark-petersen, I've force pushed the changes @bradyrx thankfully identified and this is now ready to test for the merge.

**Comment by pwolfram** [comment]
@maltrud took a look at a BGC G-case application with particles on and didn't find any problems (example below):

Screen Shot 2019-04-26 at 9 34 28 AM

@mark-petersen, this is ready to merge when you have a chance.

**Comment by bradyrx** [comment]
@pwolfram, thanks for handling this! Testing from @maltrud confirms that passive float sensors are now good for physics and biogeochem.

**Comment by pwolfram** [comment]
@mark-petersen, please let me know if there is anything else you need us to do before this gets merged-- it is ready for your final approval and merge.  Thanks!

**Comment by bradyrx** [comment]
@pwolfram, I believe that this PR still breaks if a run is restarted without all procs assigned to at least one particle. Unless you fixed that issue? @maltrud and I were going to just manually bump a particle to the empty proc(s), although that might cause load-balancing issues.

**Comment by pwolfram** [comment]
@bradyrx, please see #224, which is a separate and independent issue from this PR (which is nearly a year old).  We should engage there so this huge chunk of work can get merged without further scope creep.

**Comment by mark-petersen** [comment]
Tested in mega-merge with MPAS nightly regression suite with gnu and intel, opt and debug, on grizzly. bfb on all tests expect were expected:
```
** Running case ZISO 20km - Smoke Test
   ** FAIL (See case_outputs/ZISO_20km_-_Smoke_Test for more information)
 ** Running case Periodic Planar 20km - LIGHT particle region reset test
   ** FAIL (See case_outputs/Periodic_Planar_20km_-_LIGHT_particle_region_reset_test for more information)
 ** Running case Periodic Planar 20km - LIGHT particle time reset test
   ** FAIL (See case_outputs/Periodic_Planar_20km_-_LIGHT_particle_time_reset_test for more information)
```
See more details here, but everything looks in order.
```
/lustre/scratch4/turquoise/mpeterse/runs/nightly_ocean_model_190508_a1256fb_gr_gfortran_openmp_pr_test
```
Testing mega-merge in E3SM now. Includes coastal branch as well.

**Comment by bradyrx** [comment]
@mark-petersen @pwolfram, with this going into develop now, it might be a good time to point out my repo https://github.com/bradyrx/automate-mpas. I'll be down there next week so maybe we can talk about a path forward with this. It might be nice to get some code review here and ideas.

But basically, one just needs to alter the `config.sh` file and run `generate_e3sm_case.sh`. It'll manage every step of the E3SM/MPAS case setup and launch the run. This is built mainly to manage all of the minute steps of getting LIGHT to run properly (and is especially useful if you're running a bunch of simulations for scaling tests). The python files alter all of the streams files, etc. to set up proper supercycling, sensor modes, particle initialization, I/O, and of course base settings for MPAS.

**Comment by mark-petersen** [comment]
test of mega-merge passed all in E3SM:
```
PASS SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu RUN time=147
PASS SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_intel RUN time=397
PASS ERS_Ld9.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-knl_gnu COMPARE_base_rest
PASS PEM_Ln9.ne30_oECv3wLI.A_WCYCL1850-DIB-ISMF_CMIP6.cori-knl_gnu COMPARE_base_modpes
PASS PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-knl_gnu COMPARE_base_single_thread
```

**Comment by pwolfram** [comment]
Thanks @mark-petersen

**Pull request merged**